### PR TITLE
Implement dynamic plugin loading

### PR DESCRIPTION
**Pull request by wchill**
Plugins can now be loaded at runtime. Currently plugins will be cycled through on a fixed rotation; the logic for this can be changed if necessary.

Define the plugin directory in `config.yml`, plugin order, priority and persistence in `plugins.yml`, and plugins can provide their own config files if they need them (which should be loaded using ConfigLoader).

### CODE REVIEW

File: plugin_manager.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        while True:
+            current_time = time.time()
+            next_module = self.getNextPlugin()
+            if next_module and not next_module == current_module:
+                current_module = next_module
+                current_plugin = imp.load_source("__init__", next_module["path"]).plugin()
+                current_plugin.setTileMatrix(twilight.interface.tile_matrix)
+            while time.time() < current_time + self.config["PLUGIN_CYCLE_LENGTH"]:
+                tile_matrix = current_plugin.getNextFrame()
+                for pixel in tile_matrix:
```
Comment (modi95) [inline plugin_manager.py]: Add comment about what happens when you write to a tile which doesn't have a twilight unit

File: plugin_manager.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            for index in range(len(self.loaded_plugins)):
+                if self.loaded_plugins[index]["priority"] < plugin_obj["priority"]:
+                    self.loaded_plugins.insert(index, plugin_obj)
+                    return
+            self.loaded_plugins.append(plugin_obj)
+        else:
+            # plugin not found
+            pass
+
+    def getNextPlugin(self):
```
Comment (modi95) [inline plugin_manager.py]: Add docstring

File: plugin_manager.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+        default_config = {
+            # Plugin data, priority in range 0-9 inclusive, persistent is boolean
+            'PLUGINS': {}
+        }
+
+        config_loader.register_config('plugin', PLUGIN_FILE_NAME, default_config)
+        self.plugins = config_loader.load_config('plugin')
+        self.loaded_plugins = []
+        self.blocked = {}
```
Comment (modi95) [inline plugin_manager.py]: What is self.blocked?

File: plugin_manager.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        return None
+
+    def start(self):
+        """Begins cycling through plugins"""
+        current_module = None  # TODO: replace with default module
+        current_plugin = None
+        while True:
+            current_time = time.time()
+            next_module = self.getNextPlugin()
+            if next_module and not next_module == current_module:
```
Comment (modi95) [inline plugin_manager.py]: Please add a default module and throw an exception.

File: plugin_manager.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                current_module = next_module
+                current_plugin = imp.load_source("__init__", next_module["path"]).plugin()
+                current_plugin.setTileMatrix(twilight.interface.tile_matrix)
+            while time.time() < current_time + self.config["PLUGIN_CYCLE_LENGTH"]:
+                tile_matrix = current_plugin.getNextFrame()
+                for pixel in tile_matrix:
+                    twilight.interface.set_unit_color(pixel, tile_matrix[pixel])
+                time.sleep(0.1)
```
Comment (modi95) [inline plugin_manager.py]: There's a flag in the config file for safe timeouts. Please do not use literals.

File: plugins.yml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+# This is the central documentation for plugin settings.
+# Plugins not listed here are non-persistent and have priority 0
+
+# Plugin data, priority in range 0-9 inclusive, persistent is boolean
+PLUGINS:
+  epilepsy:
+    priority: 0
+    persistent: False
```
Comment (modi95) [inline plugins.yml:10]: What does persistent do @warut-vijit

**Review by modi95 — CHANGES_REQUESTED** [review]
Please add docstrings. And avoid using literals.

File: plugins.yml (hunk 6 above)
Comment (warut-vijit) [reply plugins.yml:10]: "Persistent" indicates whether a particular plugin should be removed from the queue after executing for one cycle or remain on the queue. An example of the latter is an alarm, which should remain on the queue by default

Comment (modi95) [reply plugins.yml:10]: Pls that to the plugins.yml header 😄

File: plugin_base.py [hunk 7]
```diff
@@ -4,10 +4,15 @@ def __init__(self):
         self.tile_matrix = []
 
     def ready(self):
+        """Specifies whether this plugin is ready to return a frame."""
         return False
 
     def setTileMatrix(self, tile_matrix):
```
Comment (modi95) [inline plugin_base.py]: We're using snek case not camel case.

File: plugin_manager.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        while True:
+            current_time = time.time()
+            next_module = self.getNextPlugin()
+            if next_module and not next_module == current_module:
+                # Load the next plugin and give it the current panel layout
+                current_module = next_module
+                current_plugin = imp.load_source("__init__", next_module["path"]).plugin()
+                current_plugin.setTileMatrix(twilight.interface.tile_matrix)
+
+            while time.time() < current_time + self.config["PLUGIN_CYCLE_LENGTH"]:
```
Comment (modi95) [inline plugin_manager.py]: I'm a bit confused about what these times are supposed to do.

File: plugin_manager.py (hunk 3 above)
Comment (warut-vijit) [reply plugin_manager.py]: Fixed, added to comments

File: plugin_manager.py (hunk 8 above)
Comment (warut-vijit) [reply plugin_manager.py]: changed variable name to be more clear

File: plugin_base.py (hunk 7 above)
Comment (warut-vijit) [reply plugin_base.py]: slippery snek cased

**Review by modi95 — APPROVED** [review]

**Pull request merged**